### PR TITLE
[DEVHAS-53] Add Go module for gitops resource generation

### PR DESCRIPTION
**Pull request by johnmcollier**
Signed-off-by: John Collier <jcollier@redhat.com>

### What does this PR do?:
Adds a separate Go module and CLI for generating GitOps resources (base, tekton, and overlay manifests). Once this is merged in, a separate PR that will run this "generator CLI" as part of a Kubernetes job will be added to HAS.

### Which issue(s)/story(ies) does this PR fixes:
https://issues.redhat.com/projects/DEVHAS/issues/DEVHAS-53?

### PR acceptance criteria:

- [x] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

  

### How to test changes / Special notes to the reviewer:
0. Log in to an OpenShift / Kubernetes cluster and `make install` the HAS CRDs (*Note:* You do not need HAS running)
1. Clone my changes and cd into gitops-generator folder.
2. Create a dummy gitops repository somewhere
3. `go build main.go`
4. `export GITHUB_TOKEN=<some-token>`
5. Create the necessary resources on your cluster:
   a. If testing base resource generation, just create a sample component
   b. If testing overlays resource generation, you'll need to create an Application, Component, Environment, Snapshot, and SEB
5. `./main --operation generate-base --component <component-cr-name> --namespace <namespace> --repoURL <repo-url>` to generate the base and tekton resources
6. `./main --operation generate-overlays --seb <seb-name> --namespace <namespace> --repoURL <repo-url>` to generate the overlay resources

### CODE REVIEW

File: gitops-generator/pkg/gitops/prepare/prepare.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2021 Red Hat, Inc.
```
Comment (maysunfaisal) [inline gitops-generator/pkg/gitops/prepare/prepare.go:2]: do we want to update copyright years 🤔

File: gitops-generator/pkg/generate/generate.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+type GitOpsGenParams struct {
+	Generator gitopsgen.Generator
+	RemoteURL string
+	Branch    string
+	Context   string
+	Token     string
+}
+
+func GenerateGitopsBase(ctx context.Context, client client.Client, component appstudiov1alpha1.Component, appFs afero.Afero, gitopsParams GitOpsGenParams) error {
```
Comment (maysunfaisal) [inline gitops-generator/pkg/generate/generate.go:28]: arent these funcs in the library? I dont see it on main anymore

Comment (johnmcollier) [reply gitops-generator/pkg/generate/generate.go:28]: The code from these is copied from the Component and ASEB (For GenerateGitopsOverlays) controllers. Basically the wrapper/setup code needed before calling the gitops generator library.

File: gitops-generator/pkg/gitops/prepare/prepare.go (hunk 1 above)
Comment (johnmcollier) [reply gitops-generator/pkg/gitops/prepare/prepare.go:2]: Oops yes

Comment (johnmcollier) [reply gitops-generator/pkg/gitops/prepare/prepare.go]: Updated

**Review by maysunfaisal — APPROVED** [review]

**Pull request closed without merging**